### ISSUE TITLE
Fullscreen mouse pointer issues- exploration

### ISSUE THREAD

**Issue by abought** [CenterForOpenScience/lookit#84]
Tracking efforts to debug fullscreen pointer issues with @kimberscott . 

### Communication
- [ ] It seems like I am seeing this event more often than Kim. I am sending a minimal testcase in which this happens very frequently for me. We will narrow down specific events or situations where this happens before directing additional debugging efforts.

### Summary
During the geometry study, the mouse pointer sometimes spontaneously reappears within the course of a single experiment. It usually hides after I move the mouse and then returns later.

There does not appear to be any single event that triggers it- eg it happens within one single run of the geometry alternation frame.

I saw this happen in Chrome 56 (Mac OS X), but not in firefox 51 (Mac OS X).

### Key discoveries so far
In order to strip this down to the absolute minimal test case, I wrote a very simple HTML page with a button that sends a single div fullscreen and hides the cursor. I still saw my cursor suddenly reappear with 1 min of idly watching the page, reproducibly. Sometimes sooner, sometimes later.

(Using dual monitors, but with chrome window as the focus)

[minimalism.html.zip](https://github.com/CenterForOpenScience/lookit/files/800350/minimalism.html.zip)

### More detailed notes
This applied to the Baby Euclid experiment on staging + code version as of 2/24/17
What has been tried:
1. Gradually stripped down the page to see if the issue was due to re-rendering of the page: progressively turned off all triangle rendering/ drawing operations (`drawTriangles`, `clearTriangles`, `presentTriangles`)
2. Examined event stream behavior to see if pointer reappearance was correlated to a single event; nothing found. Pointer reappeared before, during, and after any given geometry alternation frame. Also progressively turned off timeouts that were triggering events until the page was simply a static picture of two rectangles for 60 seconds at a clip.
3. Using chrome devtools timeline feature, tried to profile for anomalous events, but nothing presented itself. (eg no major GC, no JS or painting activity; just the same stream of network activity as for anything else) 
4. Tried with Devtools both open and closed during experiment session to see if debug mode triggered different behavior. Also restarted browser completely in case any state was preserved.
5. Investigated Chrome issue tracker ([fullscreen api](https://bugs.chromium.org/p/chromium/issues/list?can=2&q=cursor+component%3ABlink%3EFullscreen+&colspec=ID+Pri+M+Stars+ReleaseBlock+Component+Status+Owner+Summary+OS+Modified&x=m&y=releaseblock&cells=ids), [browser fullscreen mode](https://bugs.chromium.org/p/chromium/issues/list?can=2&q=cursor+component%3AUI%3EBrowser%3EFullScreen+&colspec=ID+Pri+M+Stars+ReleaseBlock+Component+Status+Owner+Summary+OS+Modified&x=m&y=releaseblock&cells=ids)). Found a number of reports of video-related fullscreen issues over several years, but no major cursor:none issues. That said, going fullscreen on a div is a much less common use case than video.

### Attempted fixes to make behavior more predictable
1. Set cursor to an image using `cursor: url(x.gif), none` in CSS; no effect.
2. Stripped down test case (see attached file) sent to Kim

### Future areas to investigate
1. The flash video capture widget is hidden but exists. We could investigate whether this is triggering any interactions
2. Investigate dimensions of flash video widget container; although we don't display it there's a possibility it could be interfering. (not clear whether flash widgets are captured in devtools profiler)
  - Will hold off on going down this path insofar as I could reproduce my issue in a page with no video and no ember whatsoever

maybe it's interfering (and it doesn't seem to be captured directly in chrome devtools profiler)

**Comment by kimberscott** [comment]
I also see the issue in the 'minimalism' demo in Chrome but not Firefox. 

In the actual study, though, I was seeing the cursor reappear in Firefox, although only after the first frame.

Given that we see *any* issue with the cursor even without video or ember, I agree it makes sense not to focus on those potential causes. Does going fullscreen on the window itself work any differently?

It looks like worst-case, we'd be dealing with [pointerlock](https://developer.mozilla.org/en-US/docs/Web/API/Element/requestPointerLock) to fix this--is that what you were mentioning earlier?

**Comment by abought** [comment]
> It looks like worst-case, we'd be dealing with pointerlock to fix this--is that what you were mentioning earlier?

Indeed. Very subtle CSS edge cases are one of the deep sea-monster territories of web development, and painfully, they can change from one browser version to the next. I was kind of hoping it was only misbehaving in chrome just for me, but...

The following things would need to be considered if we went ahead with pointer lock:

1. Compatibility (see [MDN chart](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_Lock_API#Browser_compatibility), [caniuse](http://caniuse.com/#feat=pointerlock))- it appears there is no internet explorer support at all, while Safari and Firefox made tweaks to their implementations relatively recently. This would limit the pool of users, or perhaps introduce another study variable for users on some old windows machines who weren't able to user pointer lock at all.

2. Like fullscreen access, the user would need to authorize this permission. We could be proactive about detecting access failures by explicitly listening for the [pointerlockerror](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_Lock_API#mouselocklostevent) event. We'd have to avoid making assumptions about what the user allowed, though. Each extra permission we have to ask for is an extra point of failure if they say no, but at least the spec gives us a way to know the current state.

**Comment by kimberscott** [comment]
Okay--good to know. It does sound like it'd be really nice to avoid needing to use pointerlock if at all possible since it'd introduce its own set of challenges, so it's still worth at least a bit more digging into why the cursor reappears and whether that could be prevented.

**Comment by abought** [comment]
> so it's still worth at least a bit more digging into why the cursor reappears and whether that could be prevented

Just a small update: I've been poking at this a bit more, but no luck yet. One other thing I tried this morning was to draw a div in front of the fullscreen element, so you'd be changing the cursor on that element instead of the fullscreen one. ([ref](https://lists.w3.org/Archives/Public/www-style/2010Aug/0407.html)) Unfortunately this didn't seem to affect the chrome issue.

Since this doesn't seem to be an easy quick fix: to use available time most efficiently, I've paused this item while reaching out to a few other groups who might have alternate perspective (including the `chromium-support` dev team via freenode [IRC](https://www.chromium.org/developers/irc), and [stackoverflow](http://stackoverflow.com/questions/42488941/mouse-pointer-not-staying-hidden-on-chrome-fullscreen-div)). No response yet but I'll keep monitoring.

**Comment by abought** [comment]
@kimberscott , could you let me know what OS / browser version you reproduced this on? We're working with the chrome dev team to trace this bug, but not everyone seems able to reproduce it. Nailing down minor differences may help/

https://bugs.chromium.org/p/chromium/issues/detail?id=696684&can=2&start=0&num=100&q=&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified&groupby=&sort

**Comment by kimberscott** [comment]
If it turns out the Chrome spontaneous-reappearance issue is relatively rare, it'd be helpful to figure out whether we can reproduce the study-specific problem. When I try the Baby Euclid study on Firefox, I don't see the cursor reappearing at any point during the first trial (even if I move it or click) but it *does* reliably reappear at the start of the second trial (when the second announcement audio is played, and the spinny ball reappears. Does that happen for you also?

In Chrome, the cursor doesn't hide at all upon going fullscreen (or maybe does that brief off-on flicker).

**Comment by abought** [comment]
> In Chrome, the cursor doesn't hide at all upon going fullscreen (or maybe does that brief off-on flicker).

Just an update: I've had some initial luck with a stackoverflow suggestion.  I wonder if this works for you as well?

```
#experiment-player.player-fullscreen, #experiment-player.player-fullscreen * {
    // blank cursor (1px transparent image)
    cursor: url(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7), none;
}
```

Basically, this replaces the cursor with a 1px blank image *or* no pointer. (I had done some experimenting with the cursor as an actual image without luck, but for whatever reason, inlining the image seems to do the trick. The solution seems weird, but so is the problem- I'd be curious if it helps for you.

**Comment by kimberscott** [comment]
It does!! I still see the cursor reappearing temporarily with *movement* of the mouse in Chrome, but it goes away properly in the first place, and doesn't come back on the second trial in *either* Chrome or Firefox. Do you also see the cursor not reappearing on trial 2?

**Comment by abought** [comment]
I went through the entire trial and the cursor didn't reappear, if that helps. Chrome and FF.

Currently, we stand at two people (us) who can see this problem in the simple test case, and three people (COS coworkers) who can't reproduce (Mac OS 10.10, 10.11, and 10.12). 

It's a little hard to gauge severity because this bug isn't uniformly reproducible for everyone, but encouraging this is working for our use cases. We can leave this ticket open for a few days but close if it seems resolved after further testing.

**Comment by kimberscott** [comment]
Okay, sounds good! I've added this fix to the geometry branch: https://github.com/kimberscott/exp-addons/commit/caa44884d5cc7d0e74c86c9cf2d6e9fdc3e2c67f

(I'm using f.lux which has had some sporadic reports of interfering with pointer behavior in fullscreen, so I tried uninstalling it--no difference to the behavior in Chrome.)

**Comment by abought** [comment]
Yeah, I don't use flux on my work laptop and still saw the issue.

I still get the mouse cursor issues after closing webstorm, evernote, and keybase, but my mouse pointer is behaving well today so it's hard to nail down. Single or dual monitors, `ember s` running or stopped... lots of variables! I'll follow up once this has had time to percolate a bit. :)
